Source/Initialization/WarpXInitData.cpp

```diff
@@ -306,3 +421,109 @@ WarpX::InitLevelDataFFT (int lev, Real time)
+
+       // Initialize IntVect based on the index type of multiFab
+       // 0 if cell-centered, 1 if node-centered.
+       IntVect mfx_type(AMREX_D_DECL(0,0,0));
+       IntVect mfy_type(AMREX_D_DECL(0,0,0));
+       IntVect mfz_type(AMREX_D_DECL(0,0,0));
+
+       for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+           mfx_type[idim] = mfx_IndexType.nodeCentered(idim);
+           mfy_type[idim] = mfy_IndexType.nodeCentered(idim);
+           mfz_type[idim] = mfz_IndexType.nodeCentered(idim);
+       }
+
+       amrex::ParallelFor (tbx, tby, tbz,
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                // Shift required in the x-, y-, or z- position
+                // depending on the index type of the multifab
+                Real fac_x = (1.0 - mfx_type[0]) * dx_lev[0]*0.5;
+                Real x = i*dx_lev[0] + real_box.lo(0) + fac_x;
+#if (AMREX_SPACEDIM==2)
+                Real y = 0.0;
+                Real fac_z = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real z = j*dx_lev[1] + real_box.lo(1) + fac_z;
+#else
+                Real fac_y = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real y = j*dx_lev[1] + real_box.lo(1) + fac_y;
+                Real fac_z = (1.0 - mfx_type[2]) * dx_lev[2]*0.5;
+                Real z = k*dx_lev[2] + real_box.lo(2) + fac_z;
+#endif
+                // Initialize the x-component of the field.
+                mfxfab(i,j,k) = xfield_wrap->getField(x,y,z);
+            },
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                Real fac_x = (1.0 - mfy_type[0]) * dx_lev[0]*0.5;
+                Real x = i*dx_lev[0] + real_box.lo(0) + fac_x;
+#if (AMREX_SPACEDIM==2)
+                Real y = 0.0;
+                Real fac_z = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real z = j*dx_lev[1] + real_box.lo(1) + fac_z;
+#elif (AMREX_SPACEDIM==3)
+                Real fac_y = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real y = j*dx_lev[1] + real_box.lo(1) + fac_y;
+                Real fac_z = (1.0 - mfx_type[2]) * dx_lev[2]*0.5;
+                Real z = k*dx_lev[2] + real_box.lo(2) + fac_z;
+#endif
+                // Initialize the y-component of the field.
+                mfyfab(i,j,k)  = yfield_wrap->getField(x,y,z);
+            },
+            [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                Real fac_x = (1.0 - mfz_type[0]) * dx_lev[0]*0.5;
+                Real x = i*dx_lev[0] + real_box.lo(0) + fac_x;
+#if (AMREX_SPACEDIM==2)
+                Real y = 0.0;
+                Real fac_z = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real z = j*dx_lev[1] + real_box.lo(1) + fac_z;
+#elif (AMREX_SPACEDIM==3)
+                Real fac_y = (1.0 - mfx_type[1]) * dx_lev[1]*0.5;
+                Real y = j*dx_lev[1] + real_box.lo(1) + fac_y;
+                Real fac_z = (1.0 - mfz_type[2]) * dx_lev[2]*0.5;
+                Real z = k*dx_lev[2] + real_box.lo(2) + fac_z;
+#endif
```

Comment:
Is the above similar to
```c++
#if (AMREX_SPACEDIM == 3)
            Real x = overlap_corner[0] + (iv[0]+r.x)*dx[0];
            Real y = overlap_corner[1] + (iv[1]+r.y)*dx[1];
            Real z = overlap_corner[2] + (iv[2]+r.z)*dx[2];
#else
            Real x = overlap_corner[0] + (iv[0]+r.x)*dx[0];
            Real y = 0.0;
#if   defined WARPX_DIM_XZ
            Real z = overlap_corner[1] + (iv[1]+r.y)*dx[1];
#elif defined WARPX_DIM_RZ
            // Note that for RZ, r.y will be theta                                                                                                                                                                                                                                                                            
            Real z = overlap_corner[1] + (iv[1]+r.z)*dx[1];
#endif
#endif
```
done in `PhysicalParticleContainer::AddPlasma`? Just checking if we could merge them.

Reply:
In principle they do something similar -- But AddPlasma is targeted for particles. As discussed.. we are not merging the two functions for now.